src/memory_prefetch.c

```diff
@@ -0,0 +1,401 @@
+ * 3. Walk the entries linked list until the key is found
+ *    If the key hasn't been found and the dictionary is in the middle of rehashing,
+ *    access the index on the second table and repeat step 3
+ *
+ * dictPrefetch executes the same algorithm as dictFind, but one step at a time
+ * for each key. Instead of waiting for data to be read from memory, it prefetches
+ * the data and then moves on to execute the next prefetch for another key.
+ *
+ * dicts - An array of dictionaries to prefetch data from.
+ * get_val_data_func - A callback function that dictPrefetch can invoke
+ * to bring the key's value data closer to the L1 cache as well.
+ */
+static void dictPrefetch(dict **dicts, GetValueDataFunc get_val_data_func) {
+    initBatchInfo(dicts, get_val_data_func);
+    KeyPrefetchInfo *info;
+    while ((info = getNextPrefetchInfo())) {
+        switch (info->state) {
+        case PREFETCH_BUCKET: prefetchBucket(info); break;
+        case PREFETCH_ENTRY: prefetchEntry(info); break;
+        case PREFETCH_KVOBJ: prefetchKVOject(info); break;
+        case PREFETCH_VALDATA: prefetchValueData(info); break;
+        default: serverPanic("Unknown prefetch state %d", info->state);
+        }
+    }
+}
+
+/* Helper function to get the value pointer of a kv object. */
+static void *getObjectValuePtr(const void *value) {
+    kvobj *kv = (kvobj *)value;
+    return (kv->type == OBJ_STRING && kv->encoding == OBJ_ENCODING_RAW) ? kv->ptr : NULL;
+}
+
+void resetCommandsBatch(void) {
+    if (batch == NULL) {
+        /* Handle the case where prefetching becomes enabled from disabled. */
+        if (server.prefetch_batch_max_size) prefetchCommandsBatchInit();
+        return;
+    }
+
+    batch->cur_idx = 0;
+    batch->keys_done = 0;
+    batch->key_count = 0;
+    batch->client_count = 0;
+    batch->executed_commands = 0;
+
+    /* Handle the case where the max prefetch size has been changed. */
+    if (batch->max_prefetch_size != (size_t)server.prefetch_batch_max_size * 2) {
+        onMaxBatchSizeChange();
+    }
+}
+
+/* The config of max prefetch size may be changed during running, the function
+ * can get the size when initializing the batch. */
+int getConfigPrefetchBatchSize(void) {
+    if (!batch) return 0;
+    /* We double the size when initializing the batch, so divide it by 2. */
+    return batch->max_prefetch_size / 2;
+}
+
+/* Prefetch command-related data:
+ * 1. Prefetch the command arguments allocated by the I/O thread to bring them
+ *    closer to the L1 cache.
+ * 2. Prefetch the keys and values for all commands in the current batch from
+ *    the main dictionaries. */
+void prefetchCommands(void) {
+    if (!batch) return;
+
+    /* Prefetch argv's for all clients */
+    for (size_t i = 0; i < batch->client_count; i++) {
+        client *c = batch->clients[i];
+        if (!c || c->argc <= 1) continue;
+        /* Skip prefetching first argv (cmd name) it was already looked up by
+         * the I/O thread. */
```

Comment:
Function comment mentions that goal is to bring data to L1. Then, here it says I/O thread already did the look up but on a multi core machine (most machines nowadays), iothread and main thread will be on different cores and not sharing L1. I wonder if this comment is accurate or am I missing something.

Reply:
hmm, maybe the comment is trying to say main thread will not touch argv[0].